### PR TITLE
Sparse vector query: reuse inference results during query rewriting

### PR DESCRIPTION
**Pull request by ioanatia**
Adds a `registerUniqueRewriteAction` to `QueryRewriteContext`.

A unique action can be registered to the context, along side a consumer.
When an action is registered multiple times, we register all consumers, but we make sure that the action will be executed only once.
During `executeAsyncActions`, all "unique" actions will be executed and their results will be passed to the associated consumers.
This is useful for us in the context of rewriting query builders that need inference results.

I am just changing `SparseVectorQueryBuilder` for now - the plan is to extend this for `SemanticQueryBuilder` after the CCS work stabilizes

AFAICS all actions that right now use `registerAsyncAction` could use the this pattern of `registerUniqueRewriteAction`.
So we can refactor all the calls to `registerAsyncAction` if we wanted to, to reuse this pattern, including the ones that make inference calls for knn or sparse_vector queries.
In which case `registerUniqueRewriteAction` can become `registerAsyncAction`.

### CODE REVIEW

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java [hunk 1]
```diff
@@ -321,38 +324,78 @@ static SetOnce<Map<FullyQualifiedInferenceId, InferenceResults>> getInferenceRes
...
+                    )
                 )
-            )
-            .toList();
+                .toList();
 
-        queryRewriteContext.registerAsyncAction((client, listener) -> {
-            GroupedActionListener<Tuple<FullyQualifiedInferenceId, InferenceResults>> gal = createGroupedActionListener(
-                inferenceResultsMapSupplier,
+            GroupedActionListener<Tuple<FullyQualifiedInferenceId, InferenceResults>> gal = new GroupedActionListener<>(
```
Comment (ioanatia) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java]: `InferenceRewriteAsyncAction` encapsulates a `GroupedActionListener` that can make multiple inference requests.
We could also have a `InferenceRewriteAsyncAction` per inference_id/query.
But we can always do this as a follow up.
In most cases, we can expect that a query uses a single inference ID (e.g. ELSER) - and this would take care of removing redundant inference calls for most cases (e.g. querying with the same query string multiple semantic_text fields that use the same inference_id)

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java [hunk 2]
```diff
@@ -321,38 +324,78 @@ static SetOnce<Map<FullyQualifiedInferenceId, InferenceResults>> getInferenceRes
         return inferenceResultsMapSupplier;
     }
 
+    @SuppressWarnings("unchecked")
```
Comment (ioanatia) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java]: I am not sure how we could refactor this in such a way that we don't need to do a unchecked cast.

File: server/src/main/java/org/elasticsearch/index/query/QueryRewriteAsyncAction.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+     * - Execute the action
+     * - Pass the action result to the consumers and execute the consumers
+     * - Call the listener
+     *
+     * @param client A rest client that can be used during execution.
+     * @param listener The listener that will be called after the action and consumers have been executed.
+     * @param consumers A list of consumer that expect the result of the action. It is the responsibility of the implementor to execute
+     *                  the consumers.
+     */
+    public abstract void execute(Client client, ActionListener<?> listener, List<Consumer<Object>> consumers);
```
Comment (ioanatia) [inline server/src/main/java/org/elasticsearch/index/query/QueryRewriteAsyncAction.java]: in the future, we can also see how we can pass a client for remote clusters - we could then refactor  https://github.com/elastic/elasticsearch/pull/138124 so all from the QueryRewriteContext POV all async actions are treated the same even if they make calls to remote clusters.

**Review by Mikep86 — COMMENTED** [review]
This is a great start! I like the high-level approach this lays out, having an async action that we can deduplicate and then having multiple consumers registered for it.

I think next steps are:

- Consider how to make this generic enough to also encapsulate remote cluster async actions
- Consider how we can clean up the interface to git rid of the generic object consumers. Conceptually, consumers for a shared async action should all consume the same output type, and the interface should represent that.

For the latter point, I think a solution may involve combining the action and the consumer (at some level), so that the consumer's type can be defined with respect to the action. It's not a complete solution as stated though, and it may involve some intermediate representations so that we can create the one to many relationship between an action and its consumers. I need to think about it a bit more 🤔

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java (hunk 2 above)
Comment (kderusso) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java]: We have this paradigm a lot, I think it's probably OK as long as it's well tested

File: server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java [hunk 4]
```diff
@@ -575,4 +583,17 @@ public void setTimeRangeFilterFromMillis(long timeRangeFilterFromMillis) {
...
+
+    /**
+     * Registers an async action that must be executed only once before the next rewrite round.
+     * A {@link Consumer} argument is also required.
+     * When an async action is registered multiple times, we simply collect all the consumers associated with it.
+     * After the async action is executed, all consumers associated with it will be executed and receive as argument
+     * the result of the async action.
+     */
+    public void registerUniqueRewriteAction(QueryRewriteAsyncAction<?> action, Consumer<Object> consumer) {
+        uniqueRewriteActions.computeIfAbsent(action, k -> new ArrayList<>()).add(consumer);
```
Comment (kderusso) [inline server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java:737]: Should we use `putIfAbsent` here?

**Review by kderusso — COMMENTED** [review]
Overall approach looks good so far, will defer to Mike on CCS concerns 👍

File: server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java [hunk 5]
```diff
@@ -618,4 +625,17 @@ public void setTimeRangeFilterFromMillis(long timeRangeFilterFromMillis) {
...
+    private final Map<QueryRewriteAsyncAction<?>, List<Consumer<Object>>> uniqueRewriteActions = new HashMap<>();
+
+    /**
+     * Registers an async action that must be executed only once before the next rewrite round.
+     * A {@link Consumer} argument is also required.
+     * When an async action is registered multiple times, we simply collect all the consumers associated with it.
+     * After the async action is executed, all consumers associated with it will be executed and receive as argument
+     * the result of the async action.
+     */
+    public void registerUniqueRewriteAction(QueryRewriteAsyncAction<?> action, Consumer<Object> consumer) {
```
Comment (Mikep86) [inline server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java]: I think we can get rid of the raw `Object` usage here by templating this method call:

```
public <T> void registerUniqueRewriteAction(QueryRewriteAsyncAction<T> action, Consumer<T> consumer) {
```

File: server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java (hunk 4 above)
Comment (ioanatia) [reply server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java:737]: why would we use `putIfAbsent` here? what's wrong with `computeIfAbsent`?

Comment (kderusso) [reply server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java:737]: This is fine as well

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/sparse_vector_search.yml [hunk 6]
```diff
@@ -121,6 +121,34 @@ teardown:
...
+        body:
+          query:
+            bool:
+              must:
+                - sparse_vector:
+                    field: ml.tokens
+                    inference_id: text_expansion_model
+                    query: "octopus comforter smells"
+                - sparse_vector:
+                    field: ml.tokens
```
Comment (kderusso) [inline x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/sparse_vector_search.yml]: Should we test for a different field as well?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/search/SparseVectorQueryBuilder.java [hunk 7]
```diff
@@ -271,61 +274,83 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
...
+
+        SparseInferenceRewriteAction(QueryRewriteContext queryRewriteContext, String inferenceId, String query) {
+            super(queryRewriteContext);
+            this.inferenceId = inferenceId;
+            this.query = query;
+        }
+
+        @Override
+        protected void execute(Client client, ActionListener<InferModelAction.Response> listener) {
+            // TODO: Move this class to `server` and update to use InferenceAction.Request
```
Comment (kderusso) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/search/SparseVectorQueryBuilder.java:298]: 🙃

File: server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+                    SetOnce<Boolean> hasRun = new SetOnce<>();
+                    ctx.registerUniqueRewriteAction(
+                        // we register the same action multiple times
+                        new TestQueryRewriteAsyncAction(ctx, actionLabel, execCounter, failAction && failedActionStep == actionLabel),
+                        (result) -> {
+                            // a consumer is executed only once
+                            hasRun.set(true);
+
+                            if (failConsumer && failedActionStep == actionLabel && failedConsumerLabel == consumerLabel) {
+                                throw new IllegalStateException("it's a trap!");
```
Comment (kderusso) [inline server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java]: 😅

**Review by kderusso — APPROVED** [review]
Nice work, changes LGTM

File: server/src/main/java/org/elasticsearch/index/query/QueryRewriteAsyncAction.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+     * - Pass the action result to the consumers and execute the consumers
+     * - Call the final listener
+     *
+     * @param client A rest client that can be used during execution.
+     * @param listener The listener that will be called after the action and consumers have been executed.
+     * @param consumers A list of consumer that expect the result of the action.
+     */
+    @SuppressWarnings("unchecked")
+    public void execute(Client client, ActionListener<?> listener, List<Consumer<?>> consumers) {
+        ActionListener<T> actionListener = new ActionListener<T>() {
```
Comment (Mikep86) [inline server/src/main/java/org/elasticsearch/index/query/QueryRewriteAsyncAction.java]: It would be cleaner to use `listener.delegateFailureAndWrap` here

File: server/src/main/java/org/elasticsearch/index/query/QueryRewriteAsyncAction.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        this.queryRewriteContext = queryRewriteContext;
+    }
+
+    /**
+     * The execute method will:
+     * - Execute the action using {@link #execute(Client, ActionListener)}
+     * - Pass the action result to the consumers and execute the consumers
+     * - Call the final listener
+     *
+     * @param client A rest client that can be used during execution.
```
Comment (Mikep86) [inline server/src/main/java/org/elasticsearch/index/query/QueryRewriteAsyncAction.java]: Nit: internal client is different than rest client. I point it out because both exist.

```suggestion
     * @param client An internal client that can be used during execution.
```

File: server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java [hunk 11]
```diff
@@ -618,4 +625,17 @@ public void setTimeRangeFilterFromMillis(long timeRangeFilterFromMillis) {
     public void setTrackTimeRangeFilterFrom(boolean trackTimeRangeFilterFrom) {
         this.trackTimeRangeFilterFrom = trackTimeRangeFilterFrom;
     }
+
+    private final Map<QueryRewriteAsyncAction<?>, List<Consumer<?>>> uniqueRewriteActions = new HashMap<>();
```
Comment (Mikep86) [inline server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java]: Nit: could we declare this towards the top of the class, where the rest of the member variables are?

File: server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java [hunk 12]
```diff
@@ -618,4 +625,17 @@ public void setTimeRangeFilterFromMillis(long timeRangeFilterFromMillis) {
...
+
+    /**
+     * Registers an async action that must be executed only once before the next rewrite round.
+     * A {@link Consumer} argument is also required.
+     * When an async action is registered multiple times, we simply collect all the consumers associated with it.
+     * After the async action is executed, all consumers associated with it will be executed and receive as argument
+     * the result of the async action.
+     */
+    public <T> void registerUniqueRewriteAction(QueryRewriteAsyncAction<T> action, Consumer<T> consumer) {
```
Comment (Mikep86) [inline server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java]: Nit: Perhaps `registerUniqueAsyncAction` is a more fitting name? It aligns better with the existing async action registration names.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/search/SparseVectorQueryBuilder.java [hunk 13]
```diff
@@ -271,61 +274,83 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
...
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj instanceof SparseInferenceRewriteAction == false) {
+                return false;
+            }
+            SparseInferenceRewriteAction other = (SparseInferenceRewriteAction) obj;
+            return Objects.equals(inferenceId, other.inferenceId) && Objects.equals(query, other.query);
+        }
```
Comment (Mikep86) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/search/SparseVectorQueryBuilder.java:359]: Nit: Generally we use `getClass`-based equality

File: server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+        TestRewritable testRewritable = new TestRewritable(false, false);
+
+        assertEquals(0, testRewritable.execCounter().intValue());
+
+        TestRewritable rewritten = Rewriteable.rewrite(testRewritable, ctx);
+
+        assertEquals(testRewritable, rewritten);
+
+        CountDownLatch latch = new CountDownLatch(1);
+        ctx.executeAsyncActions(new ActionListener<>() {
```
Comment (Mikep86) [inline server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java]: Nit: Use `ActionListener.wrap` here?

File: server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+                throw new AssertionError(e);
+            }
+        });
+
+        latch.await();
+
+        // we check the number of actions that were executed
+        assertEquals(rewritten.numberOfAsyncActions(), testRewritable.execCounter().intValue());
+        // we check that for each action, all consumers were executed
+        assertEquals(rewritten.actualAsyncResults(), rewritten.expectedAsyncResults());
```
Comment (Mikep86) [inline server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java]: Nit: `assertEquals` takes the expected value first. I mention it because it can lead to misleading assertion failure messages.

File: server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                latch.countDown();
+                listenerException.set(e);
+            }
+        });
+
+        latch.await();
```
Comment (Mikep86) [inline server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java]: There's some common code we can factor out into a helper method that could:

- Create the `QueryRewriteContext`
- Create the `CountDownLatch`
- Call `executeAsyncActions` and wait for the latch

The method signature could look something like `executeAsyncActions(TestRewritable testRewritable, ActionListener<Void> listener)`?

File: server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+                latch.countDown();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                throw new AssertionError(e);
+            }
+        });
+
+        latch.await();
```
Comment (Mikep86) [inline server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java]: We may want to pass a timeout to `await` so that we don't block for a long time on a failure

File: server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+        private final boolean failConsumer;
+        private final boolean failAction;
+
+        TestRewritable(boolean failConsumer, boolean failAction) {
+            this.failConsumer = failConsumer;
+            this.failAction = failAction;
+        }
+
+        @Override
+        public TestRewritable rewrite(QueryRewriteContext ctx) throws IOException {
```
Comment (Mikep86) [inline server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java:167]: Great implementation of randomized actions and consumers!

File: server/src/main/java/org/elasticsearch/index/query/QueryRewriteAsyncAction.java [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+ * {@link QueryRewriteContext#registerUniqueRewriteAction(QueryRewriteAsyncAction, Consumer)}.
+ * This is useful when we want to remove duplicate and costly async actions that take part in query rewriting, such as generating
+ * embeddings for semantic search.
+ * Since we need to determine whether an action has already been registered, we require implementors to provide implementations for
+ * {@link #hashCode()} and {@link #equals(Object)},
+ */
+public abstract class QueryRewriteAsyncAction<T> {
+    private final QueryRewriteContext queryRewriteContext;
+
+    public QueryRewriteAsyncAction(QueryRewriteContext queryRewriteContext) {
```
Comment (Mikep86) [inline server/src/main/java/org/elasticsearch/index/query/QueryRewriteAsyncAction.java]: Why does this take `QueryRewriteContext`? I don't see it used in any implementations.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/search/SparseVectorQueryBuilder.java [hunk 20]
```diff
@@ -271,61 +274,83 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
...
+                            + "] a compatible model?"
+                    );
+                }
+            }
         );
 
         return new SparseVectorQueryBuilder(this, textExpansionResultsSupplier);
     }
 
+    private static class SparseInferenceRewriteAction extends QueryRewriteAsyncAction<InferModelAction.Response> {
```
Comment (Mikep86) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/search/SparseVectorQueryBuilder.java]: We could modify this so the consumers get a `TextExpansionResults` instance and we do the validation once in the `execute` method

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/sparse_vector_search.yml [hunk 21]
```diff
@@ -121,6 +121,34 @@ teardown:
   - match: { hits.total.value: 4 }
   - match: { hits.hits.0._source.source_text: "the octopus comforter smells" }
 
+---
+"Test multiple sparse_vector to check for embeddings reuse for duplicated query and inference id":
```
Comment (Mikep86) [inline x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/sparse_vector_search.yml]: Do we really need this test? It doesn't really demonstrate that we de-duplicate the the inference calls. This same exact test would pass with no changes to `SparseVectorQueryBuilder`.

In order to test that the inference calls are de-duplicated, we would need to instrument the rewrite process, which we already do in `QueryRewriteAsyncActionTests`.

**Review by Mikep86 — COMMENTED** [review]
Nice work! I left some comments, but most of them are nits.

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/sparse_vector_search.yml (hunk 21 above)
Comment (ioanatia) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/sparse_vector_search.yml]: We don't have any yaml test that actually test duplicate inference calls for sparse vector.
Do we feel comfortable for now with just having the `QueryRewriteAsyncActionTests`?

File: server/src/main/java/org/elasticsearch/index/query/QueryRewriteAsyncAction.java [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    protected QueryRewriteContext getQueryRewriteContext() {
+        return queryRewriteContext;
+    }
+
+    /**
+     * This method will execute the async action and pass its result to the listener.
+     *
+     * @param client A rest client that can be used during execution.
```
Comment (dimitris-athanasiou) [inline server/src/main/java/org/elasticsearch/index/query/QueryRewriteAsyncAction.java]: same for `rest` -> `internal`. Just pointing it out in case it gets lost :-)

File: server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java [hunk 23]
```diff
@@ -381,18 +382,18 @@ public void registerRemoteAsyncAction(
...
+        return asyncActions.isEmpty() == false || uniqueRewriteActions.isEmpty() == false || remoteAsyncActions.isEmpty() == false;
     }
 
     /**
      * Executes all registered async actions and notifies the listener once it's done. The value that is passed to the listener is always
      * <code>null</code>. The list of registered actions is cleared once this method returns.
      */
     public void executeAsyncActions(ActionListener<Void> listener) {
-        if (asyncActions.isEmpty() && remoteAsyncActions.isEmpty()) {
+        if (asyncActions.isEmpty() && uniqueRewriteActions.isEmpty() && remoteAsyncActions.isEmpty()) {
```
Comment (dimitris-athanasiou) [inline server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java]: call `hasAsyncActions()` here instead?

File: server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java (hunk 18 above)
Comment (dimitris-athanasiou) [reply server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java:167]: ++

**Review by dimitris-athanasiou — COMMENTED** [review]
Good stuff! Added a few minor points.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/search/SparseVectorQueryBuilder.java (hunk 13 above)
Comment (ioanatia) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/search/SparseVectorQueryBuilder.java:359]: I don't think that's the general case, we have a lot of cases where we use instanceof 🙃

File: server/src/main/java/org/elasticsearch/index/query/QueryRewriteAsyncAction.java (hunk 19 above)
Comment (ioanatia) [reply server/src/main/java/org/elasticsearch/index/query/QueryRewriteAsyncAction.java]: 💩 - you are right - I think I wanted to use it such that the execute method no longer needs to pass the `client`.
we could infer the client value from `queryRewriteContext`, meaning that for remote actions, it would be up to `QueryRewriteAsyncAction` implementations to create the remote client with the right origin, as opposed to having that done in QueryRewriteContext as it is done now.
but we can follow up on that separately and for now I will remove it

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/sparse_vector_search.yml (hunk 21 above)
Comment (Mikep86) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/sparse_vector_search.yml]: Personally, i say yes. An important aspect of an effective test is an observable side-effect of the tested behavior. This test fails that check, as it will pass with the same observed behavior regardless of whether we de-duplicate the inference calls.

The current YAML tests prove that this change doesn't introduce any regressions, which I think is good enough. As stated earlier, the `QueryRewriteAsyncActionTests` demonstrate that the unique calls are de-duplicated as expected.

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/sparse_vector_search.yml (hunk 6 above)
Comment (ioanatia) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/sparse_vector_search.yml]: we removed the yaml test in the end https://github.com/elastic/elasticsearch/pull/138193#discussion_r2608241897 so I don't think we need to do this here.

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM 🚀

File: server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java [hunk 24]
```diff
@@ -21,89 +21,99 @@
...
+        };
+        rewrite(testRewritable, listener);
 
         assertTrue(listenerException.get() instanceof IllegalStateException);
         assertEquals(expectedError, listenerException.get().getMessage());
     }
 
+    private TestRewritable rewrite(TestRewritable testRewritable, ActionListener<Void> listener) throws IOException, InterruptedException {
+        QueryRewriteContext ctx = new QueryRewriteContext(null, null, null);
+        TestRewritable rewritten = Rewriteable.rewrite(testRewritable, ctx);
```
Comment (Mikep86) [inline server/src/test/java/org/elasticsearch/index/query/QueryRewriteAsyncActionTests.java:89]: Nit: It may be good to check that we have async actions queued (via `ctx.hasAsyncActions`) after this `Rewriteable.rewrite` call

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/search/SparseVectorQueryBuilder.java [hunk 25]
```diff
@@ -276,34 +276,9 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
...
-                            + inferenceResults.get(0).getWriteableName()
-                            + "]. Is ["
-                            + inferenceId
-                            + "] a compatible model?"
-                    );
-                }
-            }
-        );
+        queryRewriteContext.registerUniqueAsyncAction(new SparseInferenceRewriteAction(inferenceId, query), inferenceResponse -> {
+            textExpansionResultsSupplier.set((TextExpansionResults) inferenceResponse.getInferenceResults().getFirst());
```
Comment (Mikep86) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/search/SparseVectorQueryBuilder.java]: We can avoid the cast here by changing the type on `SparseInferenceRewriteAction` to `TextExpansionResults`

**Review by Mikep86 — APPROVED** [review]
LGTM! I think we can clean up `SparseInferenceRewriteAction` a bit before merging, but that is minor.
